File: app/routers/productos.py
from fastapi import APIRouter, Depends, Form, Request
from fastapi.responses import HTMLResponse, RedirectResponse

from app.db import pool
from app.deps import require_admin, require_dashboard
from app.templating import templates

router = APIRouter(prefix="/dashboard/productos")

SELECT_PRODUCTOS = "SELECT id, nombre, categoria, precio_venta FROM productos ORDER BY categoria, nombre"


async def _agrupados():
    productos = await pool().fetch(SELECT_PRODUCTOS)
    por_categoria: dict[str, list] = {}
    for p in productos:
        por_categoria.setdefault(p["categoria"], []).append(p)
    categorias = sorted({p["categoria"] for p in productos})
    return productos, por_categoria, categorias


@router.get("", response_class=HTMLResponse)
async def productos_page(
    request: Request, session: dict = Depends(require_dashboard), edit: int | None = None, error: str | None = None
):
    productos, por_categoria, categorias = await _agrupados()
    return templates.TemplateResponse(
        request,
        "dashboard/productos.html",
        {
            "session": session,
            "active": "productos",
            "por_categoria": por_categoria,
            "categorias": categorias,
            "editing_id": edit,
            "error": error,
        },
    )


@router.post("", response_class=HTMLResponse)
async def crear_producto(
    request: Request,
    session: dict = Depends(require_admin),
    nombre: str = Form(""),
    categoria: str = Form(""),
    precio_venta: str = Form(""),
):
    nombre, categoria = nombre.strip(), categoria.strip()
    try:
        precio = float(precio_venta)
    except ValueError:
        precio = None

    if not nombre or not categoria or precio is None:
        _, por_categoria, categorias = await _agrupados()
        return templates.TemplateResponse(
            request,
            "dashboard/productos.html",
            {
                "session": session,
                "active": "productos",
                "por_categoria": por_categoria,
                "categorias": categorias,
                "editing_id": None,
                "error": "Completa nombre, categoría y precio.",
            },
            status_code=400,
        )

    await pool().execute(
        "INSERT INTO productos (nombre, categoria, precio_venta, es_inventariable) VALUES ($1, $2, $3, false)",
        nombre,
        categoria,
        precio,
    )
    return RedirectResponse("/dashboard/productos", status_code=303)


@router.post("/{producto_id}", response_class=HTMLResponse)
async def actualizar_producto(
    request: Request,
    producto_id: int,
    session: dict = Depends(require_admin),
    nombre: str = Form(""),
    categoria: str = Form(""),
    precio_venta: str = Form(""),
):
    nombre, categoria = nombre.strip(), categoria.strip()
    try:
        precio = float(precio_venta)
    except ValueError:
        precio = None

    if not nombre or not categoria or precio is None:
        _, por_categoria, categorias = await _agrupados()
        return templates.TemplateResponse(
            request,
            "dashboard/productos.html",
            {
                "session": session,
                "active": "productos",
                "por_categoria": por_categoria,
                "categorias": categorias,
                "editing_id": producto_id,
                "error": "Datos inválidos.",
            },
            status_code=400,
        )

    await pool().execute(
        "UPDATE productos SET nombre = $1, categoria = $2, precio_venta = $3 WHERE id = $4",
        nombre,
        categoria,
        precio,
        producto_id,
    )
    return RedirectResponse("/dashboard/productos", status_code=303)


@router.post("/{producto_id}/eliminar")
async def eliminar_producto(producto_id: int, session: dict = Depends(require_admin)):
    await pool().execute("DELETE FROM productos WHERE id = $1", producto_id)
    return RedirectResponse("/dashboard/productos", status_code=303)

File: src/app/actions/productos.ts
"use server";

import { revalidatePath } from "next/cache";
import { createServerSupabaseClient } from "@/lib/supabase/server";
import { requireAdmin } from "@/lib/auth-guard";

export type ProductoState = { error?: string };

export async function crearProducto(_prev: ProductoState, formData: FormData): Promise<ProductoState> {
  await requireAdmin();

  const nombre = String(formData.get("nombre") || "").trim();
  const categoria = String(formData.get("categoria") || "").trim();
  const precio = Number(formData.get("precio_venta"));

  if (!nombre || !categoria || Number.isNaN(precio)) {
    return { error: "Completa nombre, categoría y precio." };
  }

  const supabase = createServerSupabaseClient();
  const { error } = await supabase.from("productos").insert({
    nombre,
    categoria,
    precio_venta: precio,
    es_inventariable: false,
  });

  if (error) return { error: error.message };

  revalidatePath("/dashboard/productos");
  return {};
}

export async function actualizarProducto(_prev: ProductoState, formData: FormData): Promise<ProductoState> {
  await requireAdmin();

  const id = Number(formData.get("id"));
  const nombre = String(formData.get("nombre") || "").trim();
  const categoria = String(formData.get("categoria") || "").trim();
  const precio = Number(formData.get("precio_venta"));

  if (!id || !nombre || !categoria || Number.isNaN(precio)) {
    return { error: "Datos inválidos." };
  }

  const supabase = createServerSupabaseClient();
  const { error } = await supabase
    .from("productos")
    .update({ nombre, categoria, precio_venta: precio })
    .eq("id", id);

  if (error) return { error: error.message };

  revalidatePath("/dashboard/productos");
  return {};
}

export async function eliminarProducto(id: number) {
  await requireAdmin();
  const supabase = createServerSupabaseClient();
  await supabase.from("productos").delete().eq("id", id);
  revalidatePath("/dashboard/productos");
}
